core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,439 @@
+
+import com.google.common.base.Preconditions
+
+/**
+ * CryptoInputStream decrypts data. It is not thread-safe. AES CTR mode is
+ * required in order to ensure that the plain text and cipher text have a 1:1
+ * mapping. The decryption is buffer based. The key points of the decryption
+ * are (1) calculating the counter and (2) padding through stream position:
+ * <p/>
+ * counter = base + pos/(algorithm blocksize);
+ * padding = pos%(algorithm blocksize);
+ * <p/>
+ * The underlying stream offset is maintained as state.
+ */
+class CryptoInputStream(
+    in: InputStream,
+    codecVal: CryptoCodec,
+    bufferSizeVal: Integer,
+    keyVal: Array[Byte],
+    ivVal: Array[Byte],
+    streamOffsetVal: Long)
+  extends FilterInputStream(in: InputStream) with ReadableByteChannel {
+  val oneByteBuf = new Array[Byte](1)
+  val codec = codecVal
+
+  val bufferSize = CryptoStreamUtils.checkBufferSize(codecVal, bufferSizeVal)
+  /**
+   * Input data buffer. The data starts at inBuffer.position() and ends at
+   * to inBuffer.limit().
+   */
+  val inBuffer: ByteBuffer = ByteBuffer.allocateDirect(bufferSizeVal)
+
+  /**
+   * The decrypted data buffer. The data starts at outBuffer.position() and
+   * ends at outBuffer.limit()
+   */
+  val outBuffer: ByteBuffer = ByteBuffer.allocateDirect(bufferSizeVal)
+  var streamOffset: Long = streamOffsetVal // Underlying stream offset.
+
+  /**
+   * Whether the underlying stream supports
+   * {@link org.apache.hadoop.fs.ByteBufferReadable}
+   */
+  var usingByteBufferRead: Boolean = false
+  var usingByteBufferReadInitialized: Boolean = false
+  /**
+   * Padding = pos%(algorithm blocksize) Padding is put into {@link #inBuffer}
+   * before any other data goes in. The purpose of padding is to put the input
+   * data at proper position.
+   */
+  var padding: Byte = '0'
+  var closed: Boolean = false
+  var key: Array[Byte] = keyVal.clone()
+  var initIV: Array[Byte] = ivVal.clone()
+  var iv: Array[Byte] = ivVal.clone()
+  var isReadableByteChannel: Boolean = in.isInstanceOf[ReadableByteChannel]
+
+  /** DirectBuffer pool */
+  var bufferPool: Queue[ByteBuffer] = new ConcurrentLinkedQueue[ByteBuffer]()
+  /** Decryptor pool */
+  var decryptorPool: Queue[Decryptor] = new ConcurrentLinkedQueue[Decryptor]()
+
+  var tmpBuf: Array[Byte] = null
+  var decryptor: Decryptor = getDecryptor
+  CryptoStreamUtils.checkCodec(codecVal)
+  resetStreamOffset(streamOffset)
+
+  def this(in: InputStream,
+      codec: CryptoCodec,
+      bufferSize: Integer,
+      key: Array[Byte],
+      iv: Array[Byte]) {
+    this(in, codec, bufferSize, key, iv, 0)
+  }
+
+  def this(in: InputStream,
+      codec: CryptoCodec,
+      key: Array[Byte],
+      iv: Array[Byte]) {
+    this(in, codec, CryptoStreamUtils.getBufferSize, key, iv)
+  }
+
+  def getWrappedStream(): InputStream = in
+
+  /**
+   * Decryption is buffer based.
+   * If there is data in {@link #outBuffer}, then read it out of this buffer.
```

Comment:
`@link` and friends are javadoc, and this is scaladoc; I'm not sure what's the scaladoc syntax for links, though.